da/celestia/celestia.go

```diff
@@ -285,10 +357,264 @@ func (c *DataAvailabilityLayerClient) RetrieveBatches(dataLayerHeight uint64) da
 	}
 
 	return da.ResultRetrieveBatch{
-		BaseResult: da.BaseResult{
-			Code:     da.StatusSuccess,
-			DAHeight: dataLayerHeight,
-		},
+		BaseDACheckResult: da.BaseDACheckResult{
+			Code: da.StatusSuccess},
 		Batches: batches,
 	}
 }
+
+func (c *DataAvailabilityLayerClient) CheckBatchAvailability(daMetaData *da.DASubmitMetaData) da.ResultCheckBatch {
+
+	var numShares []int
+	var indexes []int
+	var proofs []*blob.Proof
+
+	DACheckMetaData := &da.DACheckMetaData{}
+	DACheckMetaData.Height = daMetaData.Height
+	DACheckMetaData.Client = daMetaData.Client
+	DACheckMetaData.Commitments = daMetaData.Commitments
+
+	dah, err := c.getDataAvailabilityHeaders(daMetaData.Height)
+	if err != nil {
+		//Returning Data Availability header Data Root for dispute validation
+		return da.ResultCheckBatch{
+			DataAvailable: false,
+			BaseDACheckResult: da.BaseDACheckResult{
+				Code:          da.StatusUnableToGetProofs,
+				Message:       "Error getting row to data root proofs",
+				CheckMetaData: DACheckMetaData,
+			},
+		}
+	}
+	DACheckMetaData.Root = dah.Hash()
+	for i, commitment := range daMetaData.Commitments {
+		included := false
+
+		proof, err := c.getProof(daMetaData.Height, commitment)
+		if err != nil || proof == nil {
+
+			//TODO (srene): Not getting proof means there is no existing data for the namespace and the commitment (the commitment is wrong).
+			//Therefore we need to prove whether the commitment is wrong or the span does not exists.
+			//In case the span is correct it is necessary to return the data for the span and the proofs to the data root, so we can prove the data
+			//is the data for the span, and reproducing the commitment will generate a different one.
+			return da.ResultCheckBatch{
+				DataAvailable: false,
+				BaseDACheckResult: da.BaseDACheckResult{
+					Code:          da.StatusUnableToGetProofs,
+					Message:       "Error getting NMT proofs",
+					CheckMetaData: DACheckMetaData,
+				},
+			}
+		}
+
+		nmtProofs := []*nmt.Proof(*proof)
+		shares := 0
+		for i, proof := range nmtProofs {
+			if i == 0 {
+				indexes = append(indexes, proof.Start())
+			}
+			shares += proof.End() - proof.Start()
+		}
+		numShares = append(numShares, shares)
+
+		if daMetaData.Indexes != nil && daMetaData.Lengths != nil {
+			if indexes[i] != daMetaData.Indexes[i] || shares != daMetaData.Lengths[i] {
```

Comment:
how u know that indexes[i] exists?
indexes appended once, i runs over nmtProofs

Reply:
if there is no index is because there is no proof returned in previous step, and therefore the function exited before.